### PR TITLE
feat: persist block with processing status

### LINKED ISSUE #1119: [Pending blocks refactor] Persist blocks with their processing state.
The idea behind this is to remove the pending blocks genserver state.

### PR DESCRIPTION
**Pull request by Arkenan**
**Description**

- Creates a BlockInfo struct that holds a signed block, a block root, and the processing status. This helps making storing operations easier, but also avoiding to recalculate the root if the block info is passed around.
- Every block store operation now uses block_info.

Next steps: use this to replace the PendingBlocks state.

Closes #1119

### CODE REVIEW

File: lib/lambda_ethereum_consensus/store/block_db.ex [hunk 1]
```diff
@@ -76,6 +104,16 @@ defmodule LambdaEthereumConsensus.Store.BlockDb do
     Logger.info("[BlockDb] Pruning finished. #{Enum.count(slots_to_remove)} blocks removed.")
   end
 
+  defp decode_block_info(root, data) do
+    with {:d, {encoded_signed_block, status}} <- {:d, :erlang.binary_to_term(data)},
```
Comment (avilagaston9) [inline lib/lambda_ethereum_consensus/store/block_db.ex:108]: Sorry, could you explain me the purpose of the :d tag here? I don't catch it.

File: lib/lambda_ethereum_consensus/store/block_db.ex [hunk 2]
```diff
@@ -10,31 +10,59 @@ defmodule LambdaEthereumConsensus.Store.BlockDb do
...
+
+    @spec from_block(SignedBeaconBlock.t(), Types.root(), block_status()) :: t()
+    def from_block(signed_block, root, status) do
+      %__MODULE__{root: root, signed_block: signed_block, status: status}
+    end
   end
 
-  def store_block(%SignedBeaconBlock{} = signed_block, block_root) do
-    {:ok, encoded_signed_block} = Ssz.to_ssz(signed_block)
+  def store_block(_root, block_info), do: store_block(block_info)
```
Comment (avilagaston9) [inline lib/lambda_ethereum_consensus/store/block_db.ex:46]: I think we are currently not using this one.

File: lib/lambda_ethereum_consensus/store/blocks.ex [hunk 3]
```diff
@@ -32,29 +32,32 @@ defmodule LambdaEthereumConsensus.Store.Blocks do
     }
   end
 
-  @spec store_block(Types.root(), SignedBeaconBlock.t()) :: :ok
-  def store_block(block_root, signed_block), do: LRUCache.put(@table, block_root, signed_block)
+  @spec store_block(BlockInfo.t()) :: :ok
+  def store_block(block_info) do
+    # The key is included in the block, no need to pass it.
+    LRUCache.put(@table, block_info.root, block_info)
```
Comment (avilagaston9) [inline lib/lambda_ethereum_consensus/store/blocks.ex:38]: As the comment says, can't we just pass `block_info`?

File: lib/lambda_ethereum_consensus/store/block_db.ex [hunk 4]
```diff
@@ -10,31 +10,59 @@ defmodule LambdaEthereumConsensus.Store.BlockDb do
...
+
+  def store_block(%BlockInfo{} = block_info) do
```
Comment (avilagaston9) [inline lib/lambda_ethereum_consensus/store/block_db.ex:48]: What do you think about renaming this into `store_block_info`? Calling `get_block` is not the same as calling `get_block_info` in `StoreBlocks`, maybe we can continue with that rule.

File: lib/lambda_ethereum_consensus/store/block_db.ex (hunk 2 above)
Comment (Arkenan) [reply lib/lambda_ethereum_consensus/store/block_db.ex:46]: You're right! Deleted.

File: lib/lambda_ethereum_consensus/store/block_db.ex (hunk 4 above)
Comment (Arkenan) [reply lib/lambda_ethereum_consensus/store/block_db.ex:48]: Good idea! Done.

File: lib/lambda_ethereum_consensus/store/block_db.ex (hunk 1 above)
Comment (Arkenan) [reply lib/lambda_ethereum_consensus/store/block_db.ex:108]: Long story short, to identify the line, but it seems like it's a controversial technique. Forum on the topic: https://elixirforum.com/t/proper-way-of-handling-simple-booleans-in-with-statement/45556/20

I'll change it to something more readable.

**Pull request merged**